File: src/domain/services/data.fetch.service.ts
import type {
	Client,
	Collection,
	Guild,
	GuildTextBasedChannel,
	Message,
} from 'discord.js';
import { Logger } from 'fonzi2';
import { hasChannelAccess } from '../../utils/permission.utils';
import { containsURL } from '../../utils/url.utils';
import type { ChainsService } from './chains.service';

export class DataFetchService {
	private readonly MSG_LIMIT = 750000;
	private readonly MSG_FETCH_MAXERRORS = 5;
	constructor(
		private client: Client<true>,
		private chainService: ChainsService
	) {}

	async fetchAllGuildMessages(guild: Guild): Promise<string[]> {
		Logger.info(`Fetching messages in guild: ${guild.name}`);
		const fetchPromises: Promise<string[]>[] =
			// ? Get all guild channels
			Array.from(guild.channels.cache.values())
				// ? Filter to only text channels with read/write access
				.filter((channel) => hasChannelAccess(this.client.user, channel))
				// ? Start fetching in each channel
				.map((channel) =>
					this.fetchChannelMessages(channel as GuildTextBasedChannel)
				);
		const results = await Promise.all(fetchPromises);
		const messages = results.flat();
		Logger.info(
			`Fetched #green${messages.length}$ messages in guild: ${guild.name}`
		);
		return messages;
	}

	private async fetchChannelMessages(
		channel: GuildTextBasedChannel
	): Promise<string[]> {
		const load = Logger.loading(`Fetching messages in #${channel.name}...`);
		const messages: string[] = [];
		let lastMessageId: string | undefined;
		let errorCount = 0;

		while (messages.length < this.MSG_LIMIT) {
			try {
				const messageBatch = await this.getMessageBatch(channel, lastMessageId);
				if (messageBatch.length === 0) break;

				const messagesContent = messageBatch.map((msg) => msg.content);
				messages.push(...messagesContent);

				void this.chainService.updateChainState(
					channel.guildId,
					messagesContent
				);
				lastMessageId = messageBatch.at(-1)?.id;

				load.update(
					`Fetched #green${messages.length}$ messages in #${channel.name}`
				);
			} catch (error) {
				Logger.warn(
					`Message fetching error in ${channel.name} at #green${messages.length}$ messages, current error count: ${errorCount}`
				);
				if (++errorCount > this.MSG_FETCH_MAXERRORS) {
					Logger.warn(
						`Fetching error limit reached in ${channel.name} at #green${messages.length}$ messages, Error ${error}`
					);
					break;
				}
			}
		}
		load.success(
			`Fetched #green${messages.length}$ messages in #${channel.name}`
		);
		return messages;
	}

	private async getMessageBatch(
		channel: GuildTextBasedChannel,
		lastMessageId?: string
	): Promise<Message<true>[]> {
		const messageBatch = (await channel.messages.fetch({
			limit: 100,
			before: lastMessageId,
		})) as Collection<string, Message<true>>;
		const cleanMessages = Array.from(messageBatch.values()).filter(
			(msg) => msg.content.split(' ').length > 1 || containsURL(msg.content)
		);
		return cleanMessages;
	}
}

File: app/services/data-fetch.service.go
package services

import (
	"rolando/app/log"
	"rolando/app/repositories"
	"strings"
	"sync"

	"github.com/bwmarrin/discordgo"
)

type DataFetchService struct {
	Session        *discordgo.Session
	MessageLimit   int
	MaxFetchErrors int
	ChainService   *ChainsService
	messagesRepo   *repositories.MessagesRepository
}

func NewDataFetchService(session *discordgo.Session, chainService *ChainsService, messagesRepo *repositories.MessagesRepository) *DataFetchService {
	return &DataFetchService{
		Session:        session,
		MessageLimit:   750000,
		MaxFetchErrors: 5,
		ChainService:   chainService,
		messagesRepo:   messagesRepo,
	}
}

// FetchAllGuildMessages fetches messages from all accessible channels in the guild.
func (d *DataFetchService) FetchAllGuildMessages(guildID string) ([]string, error) {
	guild, err := d.Session.State.Guild(guildID)
	if err != nil {
		return nil, err
	}

	var wg sync.WaitGroup
	messageCh := make(chan []string, len(guild.Channels))
	for _, channel := range guild.Channels {
		if !d.hasChannelAccess(channel) {
			continue
		}

		wg.Add(1)
		go func(channel *discordgo.Channel) {
			defer wg.Done()
			messages, err := d.fetchChannelMessages(channel)
			if err != nil {
				log.Log.Errorf("Failed to fetch messages for channel #%s: %v", channel.Name, err)
				return
			}
			messageCh <- messages
		}(channel)
	}

	wg.Wait()
	close(messageCh)

	var allMessages []string
	for msgs := range messageCh {
		allMessages = append(allMessages, msgs...)
	}

	log.Log.Infof("Fetched %d messages in guild %s", len(allMessages), guild.Name)
	return allMessages, nil
}

func (d *DataFetchService) fetchChannelMessages(channel *discordgo.Channel) ([]string, error) {
	var messages []string
	var lastMessageID string
	errorCount := 0

	for len(messages) < d.MessageLimit {
		batch, err := d.getMessageBatch(channel.ID, lastMessageID)
		if err != nil {
			errorCount++
			if errorCount > d.MaxFetchErrors {
				log.Log.Warnf("Error limit reached for channel #%s: %v", channel.Name, err)
				break
			}
			continue
		}

		if len(batch) == 0 {
			break
		}

		for _, msg := range batch {
			messages = append(messages, msg.Content)
			// Update chain state (assumes ChainService has an UpdateChainState method)
			go d.ChainService.UpdateChainState(msg.GuildID, []string{msg.Content})

		}
		go d.messagesRepo.AddMessagesToGuild(channel.GuildID, messages)
		lastMessageID = batch[len(batch)-1].ID
	}

	log.Log.Infof("Fetched %d messages from channel #%s", len(messages), channel.Name)
	return messages, nil
}

func (d *DataFetchService) getMessageBatch(channelID, lastMessageID string) ([]*discordgo.Message, error) {

	messages, err := d.Session.ChannelMessages(channelID, 100, lastMessageID, "", "")
	if err != nil {
		return nil, err
	}

	var cleanMessages []*discordgo.Message
	for _, msg := range messages {
		if len(strings.Fields(msg.Content)) > 1 || d.containsURL(msg.Content) {
			cleanMessages = append(cleanMessages, msg)
		}
	}
	return cleanMessages, nil
}

func (d *DataFetchService) hasChannelAccess(channel *discordgo.Channel) bool {
	permissions, err := d.Session.State.UserChannelPermissions(d.Session.State.User.ID, channel.ID)
	if err != nil {
		return false
	}
	return permissions&discordgo.PermissionViewChannel != 0
}

func (d *DataFetchService) containsURL(content string) bool {
	return strings.Contains(content, "http://") || strings.Contains(content, "https://")
}
